src/state.rs

```diff
@@ -94,14 +110,45 @@ impl MintersStore {
 pub static BALANCES: Item<u128> = Item::new(PREFIX_BALANCES);
 pub struct BalancesStore {}
 impl BalancesStore {
+    fn save(store: &mut dyn Storage, account: &Addr, amount: u128) -> StdResult<()> {
+        let balances = BALANCES.add_suffix(account.as_str().as_bytes());
+        balances.save(store, &amount)
+    }
+
     pub fn load(store: &dyn Storage, account: &Addr) -> u128 {
         let balances = BALANCES.add_suffix(account.as_str().as_bytes());
         balances.load(store).unwrap_or_default()
     }
 
-    pub fn save(store: &mut dyn Storage, account: &Addr, amount: u128) -> StdResult<()> {
-        let balances = BALANCES.add_suffix(account.as_str().as_bytes());
-        balances.save(store, &amount)
+    pub fn update_balance(
+        store: &mut dyn Storage,
+        account: &Addr,
+        amount: u128,
+        decoys: Option<Vec<Addr>>,
+        entropy: Option<Binary>,
+    ) -> StdResult<()> {
+        match decoys {
+            None => Self::save(store, account, amount),
+            Some(decoys_vec) => {
+                let mut accounts_to_be_written: Vec<(Addr, u128)> = vec![];
+
+                accounts_to_be_written.push((account.clone(), amount));
+                for decoy in decoys_vec.iter() {
+                    // Please note that decoys are not always present in the DB. In this case it is ok beacuse load will return 0.
+                    accounts_to_be_written.push((decoy.clone(), Self::load(store, decoy)));
+                }
+
+                let user_entropy: [u8; SHA256_HASH_SIZE] = match entropy {
+                    None => [0u8; SHA256_HASH_SIZE],
+                    Some(e) => sha_256(&e.0),
+                };
+
+                let mut rng = Prng::new(&PrngStore::load(store)?, &user_entropy);
```

Comment:
Maybe we don't need to use the entropy stored in the contract? because it is in the user's incentive to provide a good entropy that was not used before. The fact that the user could predict the resulting shuffled vector is not important.